### PR TITLE
Make C++ compile and run under gcc-4.4

### LINKED ISSUE #1928: Test building C/C++ with GCC on Jenkins
4.4, 4.6, 4.9, ~~5.1~~ 5.3
opt, dbg.

### PR DESCRIPTION
**Pull request by vjpai**
Fixes #1928 
Basic concepts are as follows

- Do not allow range-based for or lambda

- No explicit pointer comparisons to nullptr . All pointers have an implicit conversion to bool and that should be used in comparisons

- Use GRPC_OVERRIDE and GRPC_FINAL, not override or final

- unique_ptr must be used with extreme caution in any kind of collection and cannot be used in map (this change has not yet been added to this PR)

A separate PR adds relevant documentation for C++ development information (#6865)

### CODE REVIEW

**Comment by vjpai** [comment]
Dealing with this has raised another issue, documented in #6866

**Comment by vjpai** [comment]
Fixes #6866

**Comment by vjpai** [comment]
Remaining issue: build projects still has all the C++ boringssl tests included. We need to remove this for this case since we are bypassing boringssl for gcc-4.4

**Comment by jtattermusch** [comment]
Nit: clang format is needed.
https://grpc-testing.appspot.com/job/gRPC_pull_requests/9773/config=opt,language=sanity,platform=linux/testReport/junit/(root)/tests/tools_distrib_clang_format_code_sh/

**Comment by vjpai** [comment]
Yeah, about that. clang-format dockerfile is currently broken because it depends on http://llvm.org/apt , which is temporarily down .

**Comment by ctiller** [comment]
I'd offer to send you a PR against this, but I rebooted my machine: someone
else nearby might be able to do so...

**Comment by vjpai** [comment]
The remaining issue here is that the C++ tests pull in the boringssl tests even if we're not (and can't) use boringssl. Any ideas how to deactivate those?

**Comment by vjpai** [comment]
Working on providing a mechanism to disable the tests we don't want.

**Comment by vjpai** [comment]
Fixes #6861

**Comment by vjpai** [comment]
Current status: need to separate out boringssl test libs from privatelibs_cxx in Makefile which is one of the dependences for building C++ tests . Will require changes to makefile generation or templates.

**Comment by vjpai** [comment]
Now we've got a protobuf issue. Need to see how g++4.4 handles friend templates as that seems to be the problem

**Comment by vjpai** [comment]
Ok, I need to back up and ask a meta question. Do we consider it a reasonable end game to have libgrpc++ build but not all the tests? Because trying to get the tests to build is getting unreasonable. Just as an example, std::to_string(int) does not exist in gcc-4.4; there is only std::to_string(long long), std::to_string(long long unsigned), or std::to_string(long double) and this is considered an ambiguous conversion. More static_casts, or lots of conversion of std::to_string to strstream, neither of which is useful in ay other case . Likewise, commit efa8430 should maybe be considered problematic because we're willy nilly changing maps that have unique_ptr values to vectors and eventually we will hit one that affects performance. Maybe there's some subset of tests that we can consider reasonable for running on gcc-4.4 ?

**Comment by vjpai** [comment]
As more progress has been made, I am willing to say that this should be completed. I have sent in a PR to protobuf to fix one incompatibility in that library; hopefully, that gets a chance for a merge. There is only one test that still uses lambda captures, and I'll be handling that shortly.

**Comment by vjpai** [comment]
This is ready for review, but will not actually pass g++-4.4 building unless we get the change described for protobuf above (google/protobuf#1692). When I've done that manually, this seems to work.

**Comment by vjpai** [comment]
Building but not running. Investigating.

**Comment by vjpai** [comment]
The tests are now able to run manually for g++-4.4 because of commit 4425828 . They will not be automated until the protobuf PR above gets merged or a workaround is put in place in proto.

**Comment by vjpai** [comment]
My PR to protobuf (google/protobuf#1692) was merged, so this should build just fine if we pull down the newest protobuf. @nicolasnoble - would appreciate any advice on packaging, etc. implications.

**Comment by dgquintas** [comment]
heh, it seems you are going to have to fight clang-format on this one...
```
-    : public CommonStressTest< ::grpc::testing::EchoTestService::AsyncService> {
+    : public CommonStressTest<::grpc::testing::EchoTestService::AsyncService> {
```

**Comment by vjpai** [comment]
Just drop the initial paamayim nekudotayim; it's cleaner. Dropping the reference from global namespace and starting with grpc. Made other change requested by clang-format as well.

**Comment by jtattermusch** [comment]
Our third_party/protobuf has been bumped to the version with gcc4.4 fixes: https://github.com/grpc/grpc/pull/6952

File: include/grpc++/ext/proto_server_reflection_plugin.h [hunk 1]
```diff
@@ -56,7 +56,7 @@ class ProtoServerReflectionPlugin : public ::grpc::ServerBuilderPlugin {
   bool has_sync_methods() const GRPC_OVERRIDE;
 
  private:
-  std::shared_ptr<::grpc::ProtoServerReflection> reflection_service_;
+  std::shared_ptr< ::grpc::ProtoServerReflection> reflection_service_;
```
Comment (dgquintas) [inline include/grpc++/ext/proto_server_reflection_plugin.h]: very interesting that clang-format doesn't undo this change... would we still want to s/ ::grpc/grpc/?

File: include/grpc++/impl/codegen/config.h [hunk 2]
```diff
@@ -54,6 +54,7 @@
 // nullptr was added in gcc 4.6
 #if (__GNUC__ * 100 + __GNUC_MINOR__ < 406)
 #define GRPC_CXX0X_NO_NULLPTR 1
+#define GRPC_CXX0X_LIMITED_TOSTRING 1
```
Comment (dgquintas) [inline include/grpc++/impl/codegen/config.h]: what's the rationale behind this?

File: include/grpc++/impl/codegen/core_codegen.h [hunk 3]
```diff
@@ -42,42 +42,44 @@ namespace grpc {
 /// Implementation of the core codegen interface.
```
Comment (dgquintas) [inline include/grpc++/impl/codegen/core_codegen.h]: woooops, this whole set of overrides is my bad :grimacing: sorry

File: test/cpp/end2end/async_end2end_test.cc [hunk 4]
```diff
@@ -235,8 +235,11 @@ class TestScenario {
         disable_blocking, credentials_type.c_str(), message_content.size());
   }
   bool disable_blocking;
-  const grpc::string credentials_type;
-  const grpc::string message_content;
+  // Although the below grpc::string's are logically const, we can't declare
+  // them const because of a limitation in the way old compilers (e.g., gcc-4.4)
+  // manage vector insertion using a copy constructor
```
Comment (dgquintas) [inline test/cpp/end2end/async_end2end_test.cc]: (╯°□°）╯︵ ┻━┻

File: include/grpc++/ext/proto_server_reflection_plugin.h (hunk 1 above)
Comment (dgquintas) [inline include/grpc++/ext/proto_server_reflection_plugin.h]: probably for consistency, as they've been removed from the other places as well.

File: test/cpp/interop/interop_client.h [hunk 5]
```diff
@@ -45,9 +45,9 @@ namespace grpc {
 namespace testing {
 
 // Function pointer for custom checks.
-using CheckerFn =
-    std::function<void(const InteropClientContextInspector&,
-                       const SimpleRequest*, const SimpleResponse*)>;
+typedef std::function<void(const InteropClientContextInspector&,
+                           const SimpleRequest*, const SimpleResponse*)>
+    CheckerFn;
```
Comment (dgquintas) [inline test/cpp/interop/interop_client.h]: ugh, using ... = ...; isn't supported either? v_v

**Comment by dgquintas** [comment]
LGTM, just one nit comment regarding s/ ::grpc/grpc/

File: include/grpc++/impl/codegen/config.h (hunk 2 above)
Comment (vjpai) [inline include/grpc++/impl/codegen/config.h]: gcc4.4 only supports std::to_string(long long), (long long unsigned), or (long double). So, for int and unsigned int types, there was an ambiguous conversion.

File: test/cpp/interop/interop_client.h (hunk 5 above)
Comment (vjpai) [inline test/cpp/interop/interop_client.h]: Sorry.

**Comment by vjpai** [comment]
Confirming that all tests pass when running gcc4.4 manually through docker

File: include/grpc++/ext/proto_server_reflection_plugin.h (hunk 1 above)
Comment (vjpai) [inline include/grpc++/ext/proto_server_reflection_plugin.h]: done
